src/main/java/com/arpnetworking/metrics/mad/sources/TransformingSource.java

```diff
@@ -0,0 +1,429 @@
+    }
+
+    private TransformingSource(final Builder builder) {
+        super(builder);
+        _source = builder._source;
+
+        final ImmutableMap.Builder<Pattern, List<String>> findReplaceBuilder =
+                ImmutableMap.builderWithExpectedSize(builder._findAndReplace.size());
+        for (final Map.Entry<String, ? extends List<String>> entry : builder._findAndReplace.entrySet()) {
+            findReplaceBuilder.put(Pattern.compile(entry.getKey()), ImmutableList.copyOf(entry.getValue()));
+        }
+        _findAndReplace = findReplaceBuilder.build();
+
+        _inject = builder._inject;
+        _remove = builder._remove;
+
+        _source.attach(new TransformingObserver(this, _findAndReplace, _inject, _remove));
+    }
+
+    private final Source _source;
+    private final ImmutableMap<Pattern, List<String>> _findAndReplace;
+    private final ImmutableMap<String, DimensionInjection> _inject;
+    private final ImmutableList<String> _remove;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TransformingSource.class);
+    private static final Splitter.MapSplitter TAG_SPLITTER = Splitter.on(';').omitEmptyStrings().trimResults().withKeyValueSeparator('=');
+
+    // NOTE: Package private for testing
+    /* package private */ static final class TransformingObserver implements Observer {
+
+        /* package private */ TransformingObserver(
+                final TransformingSource source,
+                final Map<Pattern, List<String>> findAndReplace,
+                final ImmutableMap<String, DimensionInjection> inject,
+                final ImmutableList<String> remove) {
+            _source = source;
+            _findAndReplace = findAndReplace;
+            _inject = inject;
+            _remove = remove;
+        }
+
+        @Override
+        public void notify(final Observable observable, final Object event) {
+            if (!(event instanceof Record)) {
+                LOGGER.error()
+                        .setMessage("Observed unsupported event")
+                        .addData("event", event)
+                        .log();
+                return;
+            }
+
+            // Merge the metrics in the record together
+            final Record record = (Record) event;
+            final Map<ImmutableMap<String, String>, Map<String, MergingMetric>> mergedMetrics = Maps.newHashMap();
+            for (final Map.Entry<String, ? extends Metric> metric : record.getMetrics().entrySet()) {
+                boolean found = false;
+                final String metricName = metric.getKey();
+                for (final Map.Entry<Pattern, List<String>> findAndReplace : _findAndReplace.entrySet()) {
+                    final Pattern metricPattern = findAndReplace.getKey();
+                    final Matcher matcher = metricPattern.matcher(metricName);
+                    if (matcher.find()) {
+                        for (final String replacement : findAndReplace.getValue()) {
```

Comment:
Can you factor this out into a separate method. Should improve readability and testability.